Add DelegatingDataFetcherExceptionHandler

rwinch opened the pull request.

bbakerman commented on src/main/java/graphql/execution/DelegatingDataFetcherExceptionHandler.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    public void setDefaultHandler(DataFetcherExceptionHandler defaultHandler) {
+        this.defaultHandler = assertNotNull(defaultHandler, () -> "defaultHandler can't be null");
+    }
```
Dont make this mutable - use a builder pattern or have constructors for it

bbakerman commented on src/main/java/graphql/execution/DelegatingDataFetcherExceptionHandler.java:
```diff
@@ -0,0 +1,84 @@
+     *
+     * @return the {@link DelegatingDataFetcherExceptionHandler} to use
+     */
+    public static DelegatingDataFetcherExceptionHandler fromThrowableTypeMapping(LinkedHashMap<Class<? extends Throwable>, DataFetcherExceptionHandler> types) {
```
A specific implementation in parameters is strange

LinkedHashMap - It should be just `Map`

bbakerman replied:
I get that it wants to be in order - but I think the caller should decide not

rwinch replied:
Order matters and `LinkedHashMap` is the only `Map` implementation that preserves order.

rwinch replied:
I've pushed an update to use a Builder which encapsulates how the mapping and order are preserved. Internally a `List<PredicateDelegateMapping>` is now used such that `PredicateDelegateMapping` is a private class with the members `Predicate<Throwable>` and `DataFetcherExceptionHandler`.

bbakerman commented on src/main/java/graphql/execution/DelegatingDataFetcherExceptionHandler.java:
```diff
@@ -0,0 +1,84 @@
+        return new DelegatingDataFetcherExceptionHandler(handlers);
+    }
+
+    private static <M extends Map<K, V>, K, V> M assertNotEmpty(M map, String message) {
```
This should go in the graphql java Assert class AND use a Supplier for the message - see that code

rwinch replied:
I've updated the code with a commit that moves this to the `Assert` class as described. However, with the introduction of the `Builder` which encapsulates the mapping in `List<PredicateDelegateMapping>` there is no need for it, so I removed it in an additional commit.

bbakerman commented on src/test/groovy/graphql/execution/DelegatingDataFetcherExceptionHandlerTest.groovy:
```diff
@@ -0,0 +1,197 @@
+
+        then: 'fail assert on validation'
+        thrown AssertException
+    }
```
I find these types of test low value - asserts are important BUT not the key "regression" point in code base.

rwinch replied:
Do you want me to remove all the tests?

rwinch replied:
I've removed all tests in `DelegatingDataFetcherExceptionHandlerTest` that are asserting `AssertException` was thrown

bbakerman commented on src/test/groovy/graphql/execution/DelegatingDataFetcherExceptionHandlerTest.groovy:
```diff
@@ -0,0 +1,197 @@
+        def actualResult = handler.onException(handlerParameters)
+
+        then: 'only the matching delegate is invoked'
+        0 * _
```
I am not sure this is testing anything??

rwinch replied:
I think `then` needs updated to be `none of the delegates are invoked`. This is verify that if none of the `delegates` match, then none of them are used (the default should be used).

rwinch replied:
I've updated `then` description to be accurate.

rwinch commented:
I believe I have addressed all of your comments. I did so in a commit per comment to make it easier to follow along with what I've done. I'm happy to squash and rebase before the merge or preserve the commits.

rstoyanchev commented:
I think having this packaged as a `DataFetcherExceptionHandler` implementation doesn't seem an effective way to enable sharing across multiple, independent exception handlers, which I think is the main goal, unless all exception handlers agree to adopt this particular mechanism. The only way I see to do enable sharing effectively is to have it built into the contract. Otherwise `DataFetcherExceptionHandler` remains nothing more than a helper, utility class for use in cases where all exception handlers are controlled by the application. 

In addition, depending on how #2356 is implemented, it may change the way this here is done.

rwinch commented:
@rstoyanchev I do agree that it will depend on the result of gh-2356, but I don't understand your other comments:

> doesn't seem an effective way to enable sharing across multiple, independent exception handlers, which I think is the main goal, unless all exception handlers agree to adopt this particular mechanism

`DelegatingDataFetcherExceptionHandler` allows a user to assemble arbitrary `DataFetcherExceptionHandler` together without the delegate `DataFetcherExceptionHandler` needing to be aware of `DelegatingDataFetcherExceptionHandler`. So I'm confused why you feel that all exception handlers need to agree to adopt this particular mechanism.

> The only way I see to do enable sharing effectively is to have it built into the contract.

Which contract are you referring to?

> Otherwise `DataFetcherExceptionHandler` remains nothing more than a helper, utility class for use in cases where all exception handlers are controlled by the application.

Why is `DataFetcherExceptionHandler` a helper/utility class or how does `DelegatingDataFetcherExceptionHandler` (this PR) change `DataFetcherExceptionHandler` into a helper/utility class where all exceptions handlers are controlled by the application?

rstoyanchev commented:
`DelegatingDataFetcherExceptionHandler` could be used to assemble arbitrary `DataFetcherExceptionHandler` implementations indeed but you need to write predicates to decide which handlers handle which exceptions, and that's knowledge best encapsulated within each `DataFetcherExceptionHandler`.

Take a scenario with two frameworks having independent `DataFetcherExceptionHandler` implementations. Ideally each would encapsulate the knowledge of what exceptions it handles and what it doesn't, but as the contract is currently, you're expected to handle all exceptions. An application could use predicates to decide which exceptions go to which handler, but then such predicates would be needed in every application to make the same decisions as far as I can see.

This is why I think the ability to have multiple exception handlers should be built into the very contracts of GraphQL Java or otherwise left as it is.

rwinch commented:
I don't think that the logic of determining when to invoke a `DataFetcherExceptionHandler` is always best embedded within the `DataFetcherExceptionHandler`. For example, there might be a `ForbiddenDataFetcherExceptionHandler` that would be invoked differently based upon the context it is being used. In a Serlvet application `ForbiddenDataFetcherExceptionHandler` might be  invoked when an `java.security.AccessControlException` is thrown. In a Spring Security application, this would happen when `org.springframework.security.access.AccessDeniedException` is thrown and there is no currently authenticated user.

For cases where the logic for both the decision and the handling logic should be in the same object the object can implement both `DelegatingDataFetcherExceptionHandler` and `Predicate<Throwable>`.

I think it's worth pointing out that this is only introducing a contract for a single implementation that should work for a large majority of use cases. It does not prevent additional implementations from having their own semantics.

rstoyanchev commented:
Right, so exception handlers have to be built with `DelegatingDataFetcherExceptionHandler` in mind (by handling specific exceptions only or by implementing Predicate) or else they wouldn't work.

bbakerman commented:
We intend to merge a "async" exception handler in #2371 .  This will affect this code.

Please have a look at tell us of any concerns
